### ISSUE TITLE
Question about AdaFace Loss Implementation

### ISSUE THREAD

**Issue by nadongjin** [leondgarse/Keras_insightface#139]
Hello,
I have a question regarding the AdaFace Loss implementation in your repository. Specifically, I’m looking at this part of the code:

[losses.py]
arcface_logits = tf.where(
    tf.cast(y_true, dtype=tf.bool),
    **tf.cos(tf.clip_by_value(tf.acos(norm_logits) - scaled_margin, self.epislon, self.cos_max_epislon)) - (self.margin + scaled_margin)**,
    norm_logits,
)

I’m wondering if the - **(self.margin + scaled_margin)** term might be making the learning process overly difficult. For instance, if scaled_margin is 0.33 and self.margin is 0.4, this results in subtracting 0.73 from the cosine value. To me, this feels like it might be excessively increasing the loss, potentially pushing the model to overcompensate. 
In my opinion, wouldn’t it make more sense to adjust this part to something like **-(self.margin + (self.margin * scaled_margin))** instead? That way, the margin adjustment would scale more proportionally with scaled_margin. I’d really appreciate your thoughts or advice on this—am I misunderstanding the intent behind this design choice?
Thank you!

**Comment by nadongjin** [comment]
I understand now. The actual comments caused some confusion. The actual scaled_margin values were between -1.0 and 0. This means that high-quality images have values close to -1.0, while low-quality images have values close to 0. In that case, your code is nice. :)

**Comment by leondgarse** [comment]
Hi there! Haven't checked Github for a long time...
- May take a try, but I'm not sure if it would be effective. One reason is that `margin` is already multiplied at [losses.py#L377](https://github.com/leondgarse/Keras_insightface/blob/master/losses.py#L377), and `scaled_margin` could be negative, which makes the adjusted value may go larger than the margin multiplied version.
- We can visualize the margin adjusting under different `scaled_margin`, but still I can't tell which could be better.
  ```py
  # Original
  pp = lambda xx, scaled_margin: tf.cos(tf.acos(xx) - scaled_margin) - (0.4 + scaled_margin)
  xx = np.arange(-1, 1, 0.01)
  for ss in np.arange(-0.4, 0.5, 0.1):
      plt.plot(xx, pp(xx, ss), label="scaled_margin={:.1f}".format(ss))
  plt.plot(xx, xx, label="xx=xx")
  plt.grid(True)
  plt.legend()
  plt.tight_layout()
  ```
  ![Image](https://github.com/user-attachments/assets/215e2e09-e6fe-4ec6-859d-84448cc691af)
  ```py
  # With the modification
  pp = lambda xx, scaled_margin: tf.cos(tf.acos(xx) - scaled_margin) - (0.4 + 0.4 * scaled_margin)
  xx = np.arange(-1, 1, 0.01)
  for ss in np.arange(-0.4, 0.5, 0.1):
      plt.plot(xx, pp(xx, ss), label="scaled_margin={:.1f}".format(ss))
  plt.plot(xx, xx, label="xx=xx")
  plt.grid(True)
  plt.legend()
  plt.tight_layout()
  ```
  ![Image](https://github.com/user-attachments/assets/c7dbe86e-3495-46be-91ff-2adc90a9c3ed)
